src/lib/Parser.hs

```diff
@@ -341,34 +359,41 @@ interfaceInstance = do
   (p, pos) <- withPos letPat
   ann <- annot uType
   keyWord WhereKW
-  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
-  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
-  return $ ULet InstanceLet (p, ann') rhs'
+  rhsEither <- mkConstructorCall ann <$> (withSrc $ (URecord . NoExt) <$>
+      interfaceRecordFields "=")
+  case rhsEither of
+    Left err  -> fail err
+    Right rhs -> let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+                 in return $ ULet InstanceLet (p, ann') rhs'
   where
     -- Here, we are traversing the type annotation to retrieve the name of
     -- the interface and generate its corresponding constructor.
-    mkRhs (WithSrc _ (UPi _ _ lrhs))      rhs = mkRhs lrhs rhs
-    mkRhs (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
+    mkConstructorCall (WithSrc _ (UPi _ arr typ)) record
+      | arr `elem` [ClassArrow, ImplicitArrow] = mkConstructorCall typ record
```

Comment:
You don't seem to be handling the case of a non-implicit arrow anywhere and I think it will currently fail with the  "Interface constructor was not applied to any type ..." message

Reply:
A fix might just be to add a line such as 
```hs
| otherwise = Left "error message here"
```
below this one